components/zcash_address/src/encoding.rs

```diff
@@ -45,50 +45,34 @@ impl FromStr for ZcashAddress {
-                    Vec::<u8>::from_base32(&data).map_err(|_| ParseError::InvalidEncoding)?;
-
-                let net = match hrp.as_str() {
-                    unified::address::Address::MAINNET => Network::Main,
-                    unified::address::Address::TESTNET => Network::Test,
-                    unified::address::Address::REGTEST => Network::Regtest,
-                    // We will not define new Bech32m address encodings.
-                    _ => {
-                        return Err(ParseError::NotZcash);
-                    }
-                };
+        // Try decoding as a unified address
+        if let Ok((net, data)) = unified::Address::decode(s) {
+            return Ok(ZcashAddress {
+                net,
+                kind: AddressKind::Unified(data),
+            });
+        }
 
-                return unified::Address::try_from_bytes(hrp.as_str(), &data[..])
-                    .map(AddressKind::Unified)
-                    .map_err(|_| ParseError::InvalidEncoding)
-                    .map(|kind| ZcashAddress { net, kind });
-            }
-            Ok((hrp, data, Variant::Bech32)) => {
-                // If we reached this point, the encoding is supposed to be valid Bech32.
-                let data =
-                    Vec::<u8>::from_base32(&data).map_err(|_| ParseError::InvalidEncoding)?;
+        // Try decoding as a Sapling address (Bech32)
+        if let Ok((hrp, data, Variant::Bech32)) = bech32::decode(s) {
+            // If we reached this point, the encoding is supposed to be valid Bech32.
+            let data = Vec::<u8>::from_base32(&data).map_err(|_| ParseError::InvalidEncoding)?;
 
-                let net = match hrp.as_str() {
-                    sapling::MAINNET => Network::Main,
-                    sapling::TESTNET => Network::Test,
-                    sapling::REGTEST => Network::Regtest,
-                    // We will not define new Bech32 address encodings.
-                    _ => {
-                        return Err(ParseError::NotZcash);
-                    }
-                };
+            let net = match hrp.as_str() {
+                sapling::MAINNET => Network::Main,
+                sapling::TESTNET => Network::Test,
+                sapling::REGTEST => Network::Regtest,
+                // We will not define new Bech32 address encodings.
+                _ => {
+                    return Err(ParseError::NotZcash);
+                }
+            };
 
-                return data[..]
-                    .try_into()
-                    .map(AddressKind::Sapling)
-                    .map_err(|_| ParseError::InvalidEncoding)
-                    .map(|kind| ZcashAddress { net, kind });
-            }
-            Err(_) => (),
+            return data[..]
+                .try_into()
+                .map(AddressKind::Sapling)
+                .map_err(|_| ParseError::InvalidEncoding)
+                .map(|kind| ZcashAddress { net, kind });
         }
 
```

Comment:
On line 99 there is a comment:
```
// If it's not valid Bech32 or Base58Check, it's not a Zcash address.
```
This is now strictly speaking incorrect and should be updated.

Reply:
It's fine as-is if you consider "Bech32" to include the "Bech32m" variant. If it's not correct now then it wasn't before.

Reply:
I wouldn't consider Bech32 to include Bech32m.

Reply:
Fixed in https://github.com/zcash/librustzcash/commit/ba3ff42e132fcbb9f91c3efabb876f643eec3078